### DISCUSSION TITLE
Q: filebeat tcp input: packet drops

### DISCUSSION THREAD

**Discussion by gebhard73** [Security-Onion-Solutions/securityonion#4796] [Unsupported Versions]
Hi,

we have a standalone installation for processing high volume proxy logs.

Config:
- about 12 million events per hour from 5 proxies
- protocol: syslog tcp 514
- currently working fine on SO 16.04.7.1 (with smaller virtual hw) for over a year now
- trying to migrate to SO 2.3.51 (new system)
- no related errors could be found in logstash, elasticsearch, kibana, filebeat, redis, ...
- new SO box: virtual system, 8 CPUs, 40 GB RAM, Grafana stats look fine

Problem:
Proxies are complaining about not getting all messages throuch once in a while (e.g. "Connection refused") for the new SO box.

Analysis:
We see dropped packets in ifconfig for the NIC, but very slowly incrementing and not related in time to the complaints of the proxies. No dropped packets in ethtool, though.

Tests:
Have tried to use UDP but comparing the numbers to the old box we see missing events on the new box.
Have tried to set up more tcp inputs for filebeat in minion file, but doesn't seem to work out of the box (no additional listeners are being created).
Have tried lowering filebeat timeout, max_message_size (AFAIK max_connections defauls to 0 which means unlimited).
So it seems to be related to the number of concurrent connection (?).

Questions:
- any idea to get an error message from any component which could give an idea what is happening?
- how could we set up an additional tcp listener w/o messing completely with SO's config?
- asking aunt Google: userland docker_proxy is said to be "slow"... any objections using "userspace-proxy": false ? Would it break something, esp. iptables config of SO?
- any other suggestion?

I'm stuck. Ideas and help greatly appreciated. Thanks!

**Comment by weslambert** [comment]
You might try adjusting the number of workers for Filebeat in the minion pillar for Filebeat:

```
filebeat:
  ls_workers: $NUMBER
```

**Reply by gebhard73** [reply]
Thanks, didn't make any difference (I've tried up to four). No errors in logfiles of filebeat (set log level to info).
Other ideas?

**Reply by TOoSmOotH** [reply]
https://discuss.elastic.co/t/recommended-number-of-workers/92168

**Reply by gebhard73** [reply]
many thanks, will test and report back with results

**Reply by gebhard73** [reply]
first results: I've tested with various combinations for:
filebeat.spool_size: xxx
output.logstash:
  bulk_max_size: xxx
  worker: xx

No change in behaviour could be observed. No error messages at all despite from the proxies. 

Other suggestions? Tnx.

**Reply by gebhard73** [reply]
spool_size seems to have been removed in release 7

**Reply by weslambert** [reply]
Other than iterating through testing with Filebeat settings, you may also try increasing the speed at which Logstash pulls from Redis (downstream).  This may assist with Filebeat otherwise waiting to send to Logstash as one Logstash pipeline waits (potentially via blocking by Redis) for the other to pull from the Redis queue. Be careful, though not to overload Redis or Elasticsearch.

https://docs.securityonion.net/en/latest/logstash.html?highlight=logstash#configuration
https://docs.securityonion.net/en/latest/redis.html#tuning

**Comment by gebhard73** [comment]
Finally I found a configuration which seems to work. Here's what helped in my case:

minions file (`/opt/so/saltstack/local/pillar/minions/$MINION_$ROLE.sls`):
```
filebeat:
  ls_bulk_max_size: 4096
  ls_workers: 8
  mem_events: 8192
  mem_flush_min_events: 2048
```

and additionally hardcoded in filebeat.yml (`cp /opt/so/saltstack/default/salt/filebeat/etc/filebeat.yml /opt/so/saltstack/local/salt/filebeat/etc/filebeat.yml`): 
add to `output.logstash` section
```
pipelining: 10
```

(for keeping track of these hardcoded changes see: https://github.com/Security-Onion-Solutions/securityonion/discussions/4441)

I don't know if the parameters make sense in this combination but this is what I've come up with after some experimenting and testing. No proxy is complaining at the moment.

Any comments or suggestions for improvements are welcomed.

What really bothers me is that there is no information when filebeat drops log lines (at least I haven't been able to configure or find anything). This is critical IMHO. This should be fixed - but I honestly don't have any idea how.

Another issue: `pipelining` should be configurable by minions file.

And again for the records: the parameter `spool_size` had been removed and doesn't do anything anymore. Even the warning about the depreciation had been removed from the code.

References:
https://www.elastic.co/guide/en/beats/filebeat/7.11/configuring-howto-filebeat.html
https://www.elastic.co/guide/en/beats/filebeat/7.11/logstash-output.html
https://www.elastic.co/guide/en/beats/filebeat/7.11/configuring-internal-queue.html

Issues:
https://github.com/Security-Onion-Solutions/securityonion/issues/4883
https://github.com/Security-Onion-Solutions/securityonion/issues/4884

**Comment by gebhard73** [comment]
FWIW - After it was fine for more than a day, we got new error messages from the proxies ... so I guess I will have to tune the parametes further.

**Comment by gebhard73** [comment]
Query the filebeat container for stats on a standalone installation:
curl -s localhost:5066/stats
curl -s localhost:5066/stats | jq '.libbeat.output.events.failed'

**Comment by gebhard73** [comment]
Here's what works as of now. I will monitor some more weeks and will update this discussion accordingly:
minions file (`/opt/so/saltstack/local/pillar/minions/$MINION_$ROLE.sls`):

```
filebeat:
  ls_bulk_max_size: 4096
  ls_workers: 12
  mem_events: 81920
  mem_flush_min_events: 2048
```

and additionally hardcoded in `filebeat.yml` (`cp /opt/so/saltstack/default/salt/filebeat/etc/filebeat.yml /opt/so/saltstack/local/salt/filebeat/etc/filebeat.yml`):
add to output.logstash section
```
pipelining: 12
```

**Reply by gebhard73** [reply]
Even with the above configuration we still get alerts from the proxy that log lines could not be sent to SO ... no error message on the SO box, no comlains from filebeat ....

Any idea how further debug? 

Again: old SO version is working fine in parallel ...
